lib/routes/builds.js

```diff
@@ -314,28 +319,105 @@ module.exports = function (app) {
+  //     // Investigate how this interacts with the CDN.
+  //     /* eslint no-unused-vars: 0*/
+  //     next();
+  //   })
 
-    //
-    // Trigger a build in carpenter through POST request, pipe the response. This
-    // route assumes valid `npm publish` like JSON as data.
-    //
-    .post(auth, function build(req, res, next) {
-      app.carpenter.build({
-        data: req.body || {}
-      }, function client(error, response) {
-        if (error) return next(error);
+  //   //
+  //   // Trigger a build in carpenter through POST request, pipe the response. This
+  //   // route assumes valid `npm publish` like JSON as data.
+  //   //
+  //   .post(auth, function build(req, res, next) {
+  //     app.carpenter.build({
+  //       data: req.body || {}
+  //     }, function client(error, response) {
+  //       if (error) return next(error);
 
-        response.pipe(res);
-      }).on('error', next);
-    });
+  //       response.pipe(res);
+  //     }).on('error', next);
+  //   });
+
+  app.routes.put('/builds/:pkg/:env?', auth, asynk(async(req, res) => {
+    // Assume no locales are supplied, so just built in en-US, and no artifactory needed
+    // Unpack the base64 string content into a proper directory of code on disk
+    // walk directory to gather files, and publish
+    // create version and package records in system
+    const payload = req.body;
+    const models = app.models;
+    const pkg = models.Package.fromPublish(payload);
+    var version = pkg.version;
+    var name = pkg.name;
+
+    const versionRecord = await manager.getVersionRecord(name, version);
+    if (versionRecord) {
+      // return early by sending error
+      return res.status(403).send({ message: 'version record ddddd already exists for this package' });
+    }
+
+    const rootPath = path.join(tmpdir(), 'pkg');
+    const publishSpec = {
+      name,
+      version,
+      env: req.params.env || 'dev'
+    };
+    const builtAssets = payload._attachments || {};
+    const numAssets = Object.keys(builtAssets).length;
+    const expectedTarName = `${name}-${version}.tgz`;
+
+    if (numAssets === 1 && builtAssets[expectedTarName]) {
+      // Case1: asset is a tarfile
+      const untarPath = path.join(rootPath, 'untar', `${pkg.name}-${pkg.version}`);
+      const distDirectory = path.join(untarPath, 'package', 'dist');
+      const content = builtAssets[expectedTarName] || '';
+      const contentData = content.data || '';
+      try {
+        await manager.untarPackage(contentData, untarPath);
+        const fileInfos = await manager.walkAndGenerateFileInfos(distDirectory);
+        await op.async(() => manager.publishWithBffs(publishSpec, fileInfos));
+        await op.async(() => manager.addVersionRecord(name, version, payload));
+        await op.async(() => manager.addPackageRecord(pkg));
+      } catch (err) {
+        return res.status(err.statusCode || 500).send({ message: err.message });
```

Comment:
we technically dont need the catches in here beacause the `asynk` wrapper will send it along to the error middleware. We just need to ensure the error middleware behaves properly with our error data structures etc.

Reply:
I think you are creating useful errors in each of them but we should ensure they are all useful.